### ISSUE TITLE
Coffman Graham with self loops 

### ISSUE THREAD

**Issue by Xupransh** [OpenKieler/elkjs#63]
It seems that the using the coffman graham layering strategy with a graph that has edges with the same source and target ids result in a stack overflow. It seems that some implementations of coffman graham only consider DAGs with no self loops, and it's possible that the elk implementation also makes that assumption.

**Comment by uruuru** [comment]
@le-cds shouldn't selfloops have been replaced by dummy constructs prior to executing the layering algorithm? From a quick glance at other layerers, they seem to specifically ignore selflopps though. 

Anyway, it should be something that can be easily added to the algorithm.

**Comment by le-cds** [comment]
The `SelfLoopPreprocessor` should remove at least certain self loops; I thought it would remove all. There are quite a few self-loop related tickets. I'll take a look at this one when I start looking at all of them (which will be soon).

**Comment by le-cds** [comment]
I tried to reproduce it to open a ticket at the ELK project, but couldn't. This graph works just fine:

```
layering.strategy: COFFMAN_GRAHAM

node N1
node N2

edge N1 -> N2
edge N2 -> N2
```

@Xupransh, do you have a minimal example which triggers the problem?

**Comment by Xupransh** [comment]
That's odd, a similar example seems to break things in my basic index.html with the bundled elk and doesn't add an edge in the live editor.

**Comment by le-cds** [comment]
Hm, since you're using elkjs, which is always a little behind ELK proper, this might then be fixed after the next update. @uruuru what do you think?

**Comment by uruuru** [comment]
It's been a while ... sorry. From a brief glance at the code I think there's a bug in the ELK's implementation in that the algorithm doesn't ignore self-loops (as it should during layering).

**Comment by le-cds** [comment]
> From a brief glance at the code I think there's a bug in the ELK's implementation in that the algorithm doesn't ignore self-loops (as it should during layering).

Alright, I'll be sure to take a look once I'm done with the new self-loop code.

**Comment by uruuru** [comment]
I tried it myself with ELK just now and it seems to work. With the new selfloop handling, do you prune away selfloops prior to layering?

**Comment by le-cds** [comment]
> With the new selfloop handling, do you prune away selfloops prior to layering?

Yes, all self loops in any case, and ports only incident to self loops under certain conditions. THe new code is not yet in master, though.

**Comment by uruuru** [comment]
Let me rephrase: with the "intermediate new" :)? 
In most layering algorithms we do something along the lines of `if (!edge.isSelfloop()) ...`. As this is not done for the Coffman-Graham I figured that might be the issue. However, with the current master it seems to be fine.

**Comment by uruuru** [comment]
I just verified that the issue seems to be fixed in elkjs `0.5.1`. 

I'd still be interested in what changed between ELK 0.4.0 and 0.5.0 for the issue to disappear - I'd presume that in 0.5.0 the selfloops are pruned prior to layering.

**Comment by le-cds** [comment]
> I'd still be interested in what changed between ELK 0.4.0 and 0.5.0 for the issue to disappear - I'd presume that in 0.5.0 the selfloops are pruned prior to layering.

You're right, that changed already with the new implementation for self loop routing in 0.5.0.
